Propublica

GokuMizuno opened the pull request:
Only added the APIs for Propublica, and OpenStates to this branch.

Test functions have been removed, as far as I can see.
Database functions are not yet included, yet, but they are the next thing to be added.

Location data has been pulled, and I can get the Senators, but the House Reps are not always accurate.  Still working on that, sorry.

jl24 commented:
Great! Please update the original description as you progress. Also, don't forget to associate this PR with an issue using a keyword.

brylie commented on project/authentication/forms.py:
```diff
@@ -12,18 +12,18 @@
 
 from api.tasks import send_email as task_send_email
-from .reserved_usernames import RESERVED_USERNAMES
+from reserved_usernames import RESERVED_USERNAMES
```
We recently had a [recent pull request](https://github.com/CiviWiki/OpenCiviWiki/pull/74) where we converted imports to use relative import (".") semantics. This was to align with the 'best practice' recommendation in the book [Two Scoops of Django](https://twoscoopspress.com/products/two-scoops-of-django-1-11).

Make sure this change is desired.

GokuMizuno replied:
At this time, yes, this change is desired.  It reverts the Python 2 to 3 changes in the 2to3 branch, which isn't yet committed.

The reversions can be recommitted, if desired.

brylie replied:
OK, cool beans. Just checking :-) Revert the changes whenever is convenient.

brylie commented:
Here is how to link this PR to an issue:
https://help.github.com/articles/closing-issues-using-keywords/

brylie commented on project/legislation/baseclass.py:
```diff
@@ -0,0 +1,85 @@
+#this goes with the congresscritters class
+from django.db import models
+
+class PClient(object):
```
What does the name 'PClient' mean? Consider giving this a more meaningful name. Also, can we separate out each API into its own client, possibly sharing a common base class?

brylie replied:
Remember to rename this class to something more meaningful. If ProPublica and OpenSecrets APIs have similar structure, create a base class for them both to inherit from.

brylie commented on project/legislation/baseclass.py:
```diff
@@ -0,0 +1,85 @@
+    PClient deals with fetching responses from eith the ProPublica API, or the OpenSecrets API
+    , and dealing with what is returned.
+
+    The httplib2 stuff is shamelessly stolen from StackOverflow
```
When using code from StackOverflow, which is quite alright to do, just include a link to the relevant answer. I.e. give attribution. Consider also upvoting both the question and answer.

brylie replied:
Might using the [requests](http://docs.python-requests.org/en/master/) library simplify this code in any way? Just wondering :-)

brylie commented on project/legislation/baseclass.py:
```diff
@@ -0,0 +1,85 @@
+        return content
+
+
+class CongresscritterDB(Models.model):
```
What does 'congresscritterDB' mean? Is there a more clear name for this database table?

brylie replied:
Database tables are more conventionally defined in separate model files. Sticking with conventions is important for maintainability and developer experience. Follow the [guidelines](https://docs.djangoproject.com/en/2.0/topics/db/models/) described in the Django documentation.

For this new feature, we know we need a database table to store the representative information. So, go ahead and create it with a standard model/migration in corresponding files.

brylie commented on project/legislation/baseclass.py:
```diff
@@ -0,0 +1,85 @@
+
+
+class CongresscritterDB(Models.model):
+    def __init__():
```
I believe we are using Celery for scheduled tasks, such as fetching API results and storing them in a SQL table. Rather than taking the approach you are using here, check out the [Celery periodic task documentation](http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html). Double check with @jl24 to see what she recommends for running a periodic task to sync these records.

brylie commented on project/legislation/baseclass.py:
```diff
@@ -0,0 +1,85 @@
+        #possible to compare, then pull only deltas?  Probably not.
+        pass
+    
+    def makeDB(self, apikey, chamber):#finish
```
Follow [documented conventions](https://docs.djangoproject.com/en/2.0/topics/db/models/) for defining this model.

brylie replied:
Please define this model in a model.py file and create a migration, as per normal Django conventions.

brylie commented on project/legislation/baseclass.py:
```diff
@@ -0,0 +1,85 @@
+        roles = models.charfield(max_length = 1023)
+        #still need to see what else to include.  Party?
+
+        #titles = r.json()
```
Clean up commented-out code.

brylie commented on project/legislation/baseclass.py:
```diff
@@ -0,0 +1,85 @@
+
+        
+
+    def updateDB(self, apikey, chamber):
```
Consider using Celery for this scheduled task.

brylie commented on project/legislation/legislatorslocation.py:
```diff
@@ -0,0 +1,35 @@
+import Model from model
```
For file names with multiple words, separate the words with underscores. This makes it easier to read the file name.

brylie commented on project/legislation/legislatorslocation.py:
```diff
@@ -0,0 +1,35 @@
+import Model from model
+
+def get_legislator_and_district(account):
```
For simplicity, lets only ever retrieve the results directly from our database.

The database will be synchronized in a separate process.

In effect, we are trying to achieve separation of concerns here, where one concern is retrieving the data from a local 'mirror' and the other concern is synchronizing that 'mirror' via a scheduled task.

brylie replied:
In terms of the model, consider making separate tables for 'legislator' and 'district' with a relationship between the two. That way, the district data are only defined in one place, rather than duplicated in each representative.

brylie commented on project/legislation/legislatorslocation.py:
```diff
@@ -0,0 +1,35 @@
+    
+    location = account.zip_code
+
+    if ((location < 10000)||(location > 100000)):
```
Consider moving this constraint to the model definition. E.g.

```py
postal_code = models.CharField(max_length=5)
```


Also, double check whether it is desirable, as the data coming from third party APIs is somewhat out of our control.

brylie commented on project/legislation/legislatorslocation.py:
```diff
@@ -0,0 +1,35 @@
+
+    return NULL;
+
+def get_legislator_ids_by_lat_long(latitude, longitude):
```
What is the goal of this function? I.e. what aspect of the legislators will be matched against the latitude and longitude coordinates?

See if we can use a spatial data model for the data and [GeoDjango distance query](https://docs.djangoproject.com/en/2.0/ref/contrib/gis/db-api/#distance-queries) to retrieve nearby results.

brylie commented on project/legislation/openstates/pyopenstates.py:
```diff
@@ -0,0 +1,450 @@
+# -*- coding: utf-8 -*-
```
Rather than storing this library in our Git repository/history, see if it can be included as a dependency and installed via pip. See the [pyopenstates documentation](https://github.com/openstates/pyopenstates#installation) for installation instructions via pip.

brylie commented on project/legislation/openstates/test_pyopenstates.py:
```diff
@@ -0,0 +1,230 @@
+#!/usr/bn/env python
```
We don't need to keep these unit tests in our repository, as they are maintained in the upstream project.

brylie commented:
I reviewed about half of the files in this PR, and requested some high and low level changes. E.g. some considerations about the data model, syncing third party API data via scheduled task(s), following conventions, file naming conventions, and including third party libraries via pip.

GokuMizuno commented:
Deleted some files as requested, updated other files, also as requested.  The pull request has been pared down to the minimum files for ProPublica/OpenStates functionality.

GokuMizuno commented on project/webapp/static/js/account.js:
```diff
@@ -382,7 +382,7 @@ cw.AccountView = BB.View.extend({
 		var _this = this;
 		$.ajax({
-			url: "http://openstates.org/api/v1/legislators/geo/?long=" + coordinates.lng + "&lat="+ coordinates.lng + "&fields=full_name", headers:{"X-APIKEY": this.ProPublicaApiKey}, dataType: "jsonp",
+			url: "http://openstates.org/api/v1/legislators/geo/?long=" + coordinates.lng + "&lat="+ coordinates.lng + "&fields=full_name", headers:{"X-APIKEY": this.OPENSTATES_API_KEY}, dataType: "jsonp",
```
Commit 909286e41f247f7243a2f7cfacf25f384a9b0c74 should resolve this issue, as it sends only the latitude/ longitude to the backend, which houses the API key.

Hope this was done correctly. :)

brylie commented on project/webapp/static/js/account.js:
```diff
@@ -382,7 +382,9 @@ cw.AccountView = BB.View.extend({
 		var _this = this;
 		$.ajax({
-			url: "http://openstates.org/api/v1/legislators/geo/?long=" + coordinates.lng + "&lat="+ coordinates.lng + "&fields=full_name", headers:{"X-APIKEY": this.OPENSTATES_API_KEY}, dataType: "jsonp",
+			url: "~/legislation/openstates/pyopenstates.py",
```
There is no need for a `.py` suffix for this URL.

Make sure you are actively studying and following Django conventions, such as using [urlpatterns](https://docs.djangoproject.com/en/2.1/intro/tutorial03/) to map functions to routes.

tractaylor commented:
hey Clint
Sorry about that.
Attached is the second page.
Trac

      From: Brylie Christopher Oxley <notifications@github.com>
 To: CiviWiki/OpenCiviWiki <OpenCiviWiki@noreply.github.com> 
Cc: Subscribed <subscribed@noreply.github.com>
 Sent: Tuesday, August 21, 2018 5:15 AM
 Subject: Re: [CiviWiki/OpenCiviWiki] Propublica (#214)
   
@brylie requested changes on this pull request.In project/legislation/baseclass.py:> @@ -0,0 +1,85 @@
+"""
+The base class from which all ProPublica and OpenSecrets functions derive their power,
+from the base client to interacting with theirrespective APIs and/or django databases.
+"""
+
+import .json
+import .logging
+import .httplib2
+#this goes with the congresscritters class
+from django.db import models
+
+class PClient(object):
Remember to rename this class to something more meaningful. If ProPublica and OpenSecrets APIs have similar structure, create a base class for them both to inherit from.—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or mute the thread.

brylie commented on GDPR_stuff.txt:
```diff
@@ -0,0 +1,28 @@
+The GDPR itself contains 11 chapters and 91 articles. The following are some of the chapters and articles that have the greatest potential impact on security operations:
```
It seems like the GDPR branch got merged into the Propublica branch. It might be worth just starting a new branch for this task, and then cherry picking the commits/changes you actually want to keep in this branch. That way, you don't have to fight against the conflicts and revision history.

brylie commented on project/api/models/category.py:
```diff
@@ -6,3 +6,9 @@ class Category(models.Model):
+    def __str__(self):
+        return self.name
+
+    def __unicode__(self):
```
Django 2.x and Python 3.x do not need the `__unicode__` method any longer, as strings are now unicode by default.

brylie commented on project/api/models/civi.py:
```diff
@@ -90,35 +92,36 @@ class Civi(models.Model):
+    def __str__(self):
+        return self.title
+
+    def __unicode__(self):
```
We shouldn't need the `__unicode__` method any more.

brylie commented on project/api/models/civi.py:
```diff
@@ -151,20 +154,20 @@ def score(self, request_acct_id=None):
         if request_acct_id:
             account = Account.objects.get(id=request_acct_id)
-            y = (1 if self.author in account.following.all().values_list('id', flat=True) else 0)
+            scores_sum = (1 if self.author in account.following.all().values_list('id', flat=True) else 0)
```
Good work making this code more readable.

brylie commented on project/api/models/thread.py:
```diff
@@ -83,6 +83,12 @@ class Thread(models.Model):
+    def __str__(self):
+        return self.title
+
+    def __unicode__(self):
```
Make sure we need the `__unicode__` method. Otherwise, remove it.

brylie commented on project/api/serializers.py:
```diff
@@ -1,4 +1,5 @@
 from rest_framework import serializers
+<<<<<<< HEAD
```
Clean up conflict resolution artifacts.

brylie commented on project/api/serializers.py:
```diff
@@ -70,15 +71,262 @@ class Meta:
+
+
+class CategoryListSerializer(serializers.ModelSerializer):
+>>>>>>> f5feb83f8f912194e913b0931668541112143eb6
```
Clean up conflict resolution artifacts.

brylie commented on project/api/views.py:
```diff
@@ -2,6 +2,7 @@
 RESTful API View Definitions
 
 """
+<<<<<<< HEAD
```
Clean up conflict resolution artifacts.

brylie commented on project/api/views.py:
```diff
@@ -28,14 +29,50 @@
     return Response({
         'categories': reverse('category-list', request=request, format=format),
         'threads': reverse('thread-list', request=request, format=format),
+<<<<<<< HEAD
```
Clean up.

brylie commented:
Not to be too difficult, but I think this PR needs a reboot. There seems to be a struggle resolving conflicts (I know what it's like, as I've gone through the struggle  many times).

A simple way out is to start a new branch, from master. Then, apply any desired changes in the new branch.

GokuMizuno commented:
Joy.  I was trying to manage conflicts.  :)

I will reboot the branch.

On Mon, Sep 17, 2018, 08:53 Brylie Christopher Oxley <
notifications@github.com> wrote:

> Not to be too difficult, but I think this PR needs a reboot. There seems
> to be a struggle resolving conflicts (I know what it's like, as I've gone
> through the struggle many times).
>
> A simple way out is to start a new branch, from master. Then, apply any
> desired changes in the new branch.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/CiviWiki/OpenCiviWiki/pull/214#issuecomment-421997747>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGgLBRJc0-uy8-LKRFEVezCA8i1okcfiks5ub5eZgaJpZM4VlQXB>
> .
>

brylie commented:
@GokuMizuno consider also breaking this task into two separate efforts:

1. Create a standard Django model to store Propublica data, with a daily job to update the data
2. Ability to link records in Propublica table to Civis, via a relationship field and UI elements

Tackling these as separate tasks can help you manage the complexity of this epic. Smaller changesets are also easier to review and tend not to get as far out of sync.

brylie commented:
Hei Goku,
I notice you are still working on this branch. The commit history in the current branch is a bit 'tangled'.

Please create a *fresh* branch from master, and then commit your desired changes to that branch.

brylie commented:
I believe we should close this PR, but not delete the branch. That way, our PR list is a bit cleaner, and we can still use code from this branch in PR #220
